pkg/plugins/gateway/algorithms/vtc/vtc_basic.go

```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2024 The Aibrix Team.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vtc
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/vllm-project/aibrix/pkg/cache"
+	"github.com/vllm-project/aibrix/pkg/metrics"
+	"github.com/vllm-project/aibrix/pkg/types"
+	"github.com/vllm-project/aibrix/pkg/utils"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+)
+
+const (
+	defaultFairnessWeight    = 0.5
+	defaultUtilizationWeight = 0.5
+	defaultMaxPodLoad        = 100.0
+	defaultInputTokenWeight  = 1.0
+	defaultOutputTokenWeight = 2.0
+)
+
+const (
+	VTC_FAIRNESS_WEIGHT     = "AIBRIX_ROUTER_VTC_BASIC_FAIRNESS_WEIGHT"
+	VTC_UTILIZATION_WEIGHT  = "AIBRIX_ROUTER_VTC_BASIC_UTILIZATION_WEIGHT"
+	VTC_MAX_POD_LOAD        = "AIBRIX_ROUTER_VTC_BASIC_MAX_POD_LOAD"
+	VTC_INPUT_TOKEN_WEIGHT  = "AIBRIX_ROUTER_VTC_BASIC_INPUT_TOKEN_WEIGHT"
+	VTC_OUTPUT_TOKEN_WEIGHT = "AIBRIX_ROUTER_VTC_BASIC_OUTPUT_TOKEN_WEIGHT"
+)
+
+var (
+	fairnessWeight    = utils.LoadEnvFloat(VTC_FAIRNESS_WEIGHT, defaultFairnessWeight)
+	utilizationWeight = utils.LoadEnvFloat(VTC_UTILIZATION_WEIGHT, defaultUtilizationWeight)
+	maxPodLoad        = utils.LoadEnvFloat(VTC_MAX_POD_LOAD, defaultMaxPodLoad)
+	inputTokenWeight  = utils.LoadEnvFloat(VTC_INPUT_TOKEN_WEIGHT, defaultInputTokenWeight)
+	outputTokenWeight = utils.LoadEnvFloat(VTC_OUTPUT_TOKEN_WEIGHT, defaultOutputTokenWeight)
+)
+
+// BasicVTCRouter implements the VTC routing algorithm
+type BasicVTCRouter struct {
+	cache          cache.Cache
+	tokenTracker   TokenTracker
+	tokenEstimator TokenEstimator
+	config         *VTCConfig
+}
+
+// NewBasicVTCRouter creates a new BasicVTCRouter with the provided token tracker and estimator
+func NewBasicVTCRouter(tokenTracker TokenTracker, tokenEstimator TokenEstimator, config *VTCConfig) (*BasicVTCRouter, error) {
+	c, err := cache.Get()
+	if err != nil {
+		klog.Error("fail to get cache store in basic-vtc router")
+		return nil, err
+	}
+
+	return &BasicVTCRouter{
+		cache:          c,
+		tokenTracker:   tokenTracker,
+		tokenEstimator: tokenEstimator,
+		config:         config,
+	}, nil
+}
+
+// Route implements the VTC routing algorithm
+func (r *BasicVTCRouter) Route(ctx *types.RoutingContext, pods types.PodList) (string, error) {
+	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
```

Comment:
do we want to re-seeding on each call here or provide a package level `rand.Source`?

Reply:
Removed explicit seeding for consistency with other routing algorithms that use `SelectRandomPod()` and call `rand.Intn` directly. Also Go 1.20+, the global random number generator is auto-seeded at startup, so manual seeding can be removed from here.